fuel-vm/src/checked_transaction.rs

```diff
@@ -128,25 +151,90 @@ impl<Tx: IntoChecked> Borrow<Tx> for Checked<Tx> {
     }
 }
 
-pub trait IntoChecked: Checkable + Sized {
+/// Performs checks for a transaction
+pub trait IntoChecked: FormatValidityChecks + Sized {
     /// Metadata produced during the check.
     type Metadata: Sized;
 
     /// Returns transaction that passed all `Checks`.
-    fn into_checked(self, block_height: Word, params: &ConsensusParameters) -> Result<Checked<Self>, CheckError> {
-        self.into_checked_basic(block_height, params)?.check_signatures()
+    fn into_checked(
+        self,
+        block_height: Word,
+        params: &ConsensusParameters,
+        gas_costs: GasCosts,
+    ) -> Result<Checked<Self>, CheckError>
+    where
+        Checked<Self>: CheckPredicates,
+    {
+        self.into_checked_basic(block_height, params)?
+            .check_signatures()?
+            .check_predicates(params, gas_costs)
     }
 
     /// Returns transaction that passed only `Checks::Basic`.
     fn into_checked_basic(self, block_height: Word, params: &ConsensusParameters) -> Result<Checked<Self>, CheckError>;
 }
 
+/// Performs predicate verification for a transaction
+pub trait CheckPredicates {
+    /// Define predicate verification logic (if any)
+    fn check_predicates(self, params: &ConsensusParameters, gas_costs: GasCosts) -> Result<Self, CheckError>
+    where
+        Self: Sized;
+}
+
+impl<Tx: IntoChecked + ExecutableTransaction + fuel_tx::field::GasLimit> CheckPredicates for Checked<Tx>
```

Comment:
Why do you need the `fuel_tx::field::GasLimit` constrain here?

Reply:
`ExecutableTransaction` contains all constrains(including `IntoChecked`)

Reply:
`Interpreter::check_predicates` requires that, since it needs the gas limit of the tx now.

Reply:
`ExecutableTransaction` has superfruit `Chargeable` that contains a `limit` function.

Reply:
It's only named `limit`? Weird name choice. Anyway, I changed to the code to use that instead.